backend/browseruserinput.go

```diff
@@ -0,0 +1,132 @@
+package backend
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+// The game engine calls ChooseCenterCard(), ChoosePlayer() or
+// DoesChoosePlayerInsteadOfCenter(). These methods are request for input.
+// These methods set BrowserUserInput.expecting to indicate that the game
+// engine is expecting a message of the corresponding message type. When the
+// user polls, he will see that the game engine is expecting the corresponding
+// message type from him and fill the message chan.
+//
+// I find the mix of locks and channels a bit confusing. The lock is necessary
+// to protect the "expecting" state. "expecting" needs to be state instead of a
+// channel/message because the browser may poll multiple times. If "expecting"
+// were a message in a channel it would be "used up" on the first poll.
+// Consider an alternate architecture where each player has its own goroutine
+// and instead of protecting the expecting state with a lock the game engine
+// sends a message to the relevant player's goroutine that it's expecting a
+// message.
+
+// Message types
+const ChooseCenterCardMsgType string = "choose-center-card"
+const ChoosePlayerMsgType string = "choose-player"
+const ChoosePlayerInsteadOfCenterMsgType string = "choose-player-instead-of-center"
+
+type BrowserUserInput struct {
+	// lock protects the expecting variable. So that writes to the
+	// variable are read atomically
+	lock sync.RWMutex
+
+	// message takes a message from goroutine for the user request and
+	// gives it to the game engine
+	message chan interface{}
+
+	// expecting is the type of message the game engine is expecting from this
+	// player right now, or "" if the game engine is not expecting a
+	// message from this player
+	expecting string
+}
+
+func NewBrowserUserInput() *BrowserUserInput {
+	return &BrowserUserInput{
+		message: make(chan interface{}, 0),
+	}
+}
+
+func (input *BrowserUserInput) ChooseCenterCard(string) int {
+	input.lock.Lock()
+	input.expecting = ChooseCenterCardMsgType
+	input.lock.Unlock()
+
+	card := <-input.message
```

Comment:
we should add a timeout to these pulls

Reply:
If you do you should probably you'll need to return the appropriate random action.

Reply:
https://github.com/zingales/theresheep/issues/7